nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp

```diff
@@ -140,6 +145,12 @@ class BtActionNode : public BT::ActionNodeBase
       // user defined callback. May modify the value of "goal_updated_"
       on_wait_for_result();
 
+      // Check if the navigation goal has been updated
+      if (config().blackboard->get<bool>("goal_preempted")) {
+        config().blackboard->set("goal_preempted", false);
+        on_goal_preempted();
```

Comment:
In general, can't recovery behaviors, while being preempted, be thought of as being cancelled? Since the preemption of the navigation request should effectively stop all recoveries, the difference between a cancellation and a preemption isn't really anything.

If that's true, then I think we can directly cancel or preempt the goal at this point. What you did instead looks like sending a new null goal that will immediately success in its current pose. I think `async_cancel_goal` is the right answer in this case. Then we can remove the `on_goal_preempted` methods from each implementation. Their respective recovery server implementations just need to handle the cancellation cleanly.

Reply:
My idea with this was to allow all action nodes to do something on preemption.
I think it also makes sense since the BtNavigator does not know which node is being executed at the moment, and if the preemption flag is not cleared after the preemption is processed, the recovery behaviours would be cancelled when executed after a preemption is done, no matter on which node.
I thought of this as a simple solution where whatever node is being executed, it is notified of the preemption and clears the preemption flag. Then the on_goal_preempted call can be overriten on each specific action node to trigger a different behavior if necessary. 
What do you think?

Reply:
> allow all action nodes to do something on preemption

I think that sounds OK as long as there's actually a node that needs it. But I don't think that should be misconstrued with actually doing the preemption. If we're being asked to preempt or cancel, we should preempt or cancel, but then a separate method to also do something before or after sounds OK. 

You could also just have the BT Action Node just preempt or cancel the servers running. You effectively do that by just setting a null goal that returns immediately, but really it should be dealt with in the server instance itself. That code also impacts the controller and planner and isn't rational for those cases.

Reply:
> But I don't think that should be misconstrued with actually doing the preemption

Maybe we can then change the call from on_goal_preempted to on_navigation_goal_preempted or something like that? I understand that it is quite confusing to talk about preemption of the server instances (local to each server's behaviour) and the navigation goal (a global goal).

> That code also impacts the controller and planner and isn't rational for those cases.

I believe that by using a similar structure as we do with on_tick or on_wait_for_result, we should not affect any node that does not define a specific behaviour for on_goal_preempted, except from the extra management of the preemption flag. This shouldn't add too much overhead, but could it be an issue?

> if the navigation request was preempted or cancelled, the respective implementations should handle and communicate that to their servers.

The existing implementation handles the cancel on the halt function of the BT Action Node, effectively sending a cancel request to their respective servers. The problem with the halt is that it recursively stops the activity of all nodes in the tree. The reason to not do a similar thing on a preemption is that as you mention, the preemption shouldn't affect all nodes equally.
The intention of adding a new function when a navigation preemption is recieved is to allow the nodes to do something about it. This is at the moment only interesting for the recoveries, but it opens new chances for other action nodes to change their behaviour when the global goal changes.


> For the recoveries, that means push along the preempt / cancellation request for it to handle (which they should then just immediately return)

If I understand you correctly, you would like to have the action servers handle the global goal preemption on their own. To do so, we would need to add a new callback on the action server to handle the preemptions. I could do that by subclassing the action server to a recovery server maybe, but I believe it won't make much sense to add that callback on the Simple Action Server.

Reply:
I think all these functions make things more confusing. This should be relatively simple. You have a behavior tree spinning and when you call a cancel or preempt at least 1 BT node is active (probably multiple, the controller which is always running basically and potentially then a recovery or planner or something). My understanding of this at the moment is that if you hit cancel, all of the BT action Node derived classes (spin, wait, backup, planner, controller) will stop (via a async_cancel_goal in halt). 

So when the parent navigation attempt requests a preemption, what its asking for is to stop what you're doing and start proceeding to the next goal smoothly. That means the controller should keep going for the short period of time until the new plan is handed to it. that means the planner, if its doing something, should immediately stop and be given a new  goal to compute (that could be done via a preemption) and the recoveries should never be immediately recalled until at that point a planner or controller failure has occurred. 

Therefore, I think the `on_preemption()` for the recoveries BT nodes should do an async cancel and the planner should be given a new goal. So I think the function you've made, if you would like to keep that to implement preemption for servers, should have an implementation of cancels for the recoveries, preempt with the updated goal for the planners, and nothing (?) for the controllers. In that way, you should probably make a BT-specific recoveries node that derives from the `BTActionNode` that the recovery actions implement, so that there's not code duplication with 3+ identical `on_preemption()`s. 

You shouldn't touch `halt` that is correct. You also shouldn't need to change anything in the simple action server. You just need to check for preemption as well as cancel and then stop processing (which seems to be laid out for you as well [here](https://github.com/ros-planning/navigation2/blob/66543509618a8100c92bb22eb029efb77967ca43/nav2_recoveries/include/nav2_recoveries/recovery.hpp#L165))

Reply:
> This should be relatively simple.

I completely agree, it should be simple. However the behaviour of the behaviour tree itself is preety complex unfortunately :)

> So when the parent navigation attempt requests a preemption, what its asking for is to stop what you're doing and start proceeding to the next goal smoothly. That means the controller should keep going for the short period of time until the new plan is handed to it. that means the planner, if its doing something, should immediately stop and be given a new goal to compute (that could be done via a preemption, also done for you, no changes required here) 

Please, correct me if I'm wrong, but I think that in the current implementation, this works thanks to the tree structure used, not because the action nodes are notified of a preemption:
* The controller node does never see an update on the global goal, since it's only caring about the navigation trajectories, generated by the planner.
* The planner action node (compute_path_to_pose_action.cpp) does not know about this change in the goal either. This means that the goal update can only work when using this action node with a decorator like the RateController, which sets the action back to idle after if finds a path (which means, the planner recieves a new goal on each update of the RateController, in the case of our examples, every second).
* It is true that we need a change in the behaviour of the recoveries.

I think we should only focus on the recoveries for the moment, but we should look for a solution that is as open as possible.

> Its also not really a general solution to just terminating all recoveries when a new navigation request comes in, which I think is expected behavior.

I completely agree, I will be happy to change it to a cancel.

---
To solve this issue with the recoveries, we should find a way to detect in the recovery nodes wether the goal has changed during their execution. I am very happy to get rid of the blackboard flag I added to check the preemption. I realised that this doesn't make much sense, since some nodes might be active at the same time.

The best would be to have a signal or something that can notify all running nodes of a change in the navigation goal. Do you know if there is something able to do so in the behaviour tree?

What about changing this code with one of these solutions:
* Creating a new action node that can be subclassed from on all recoveries including a function that can: 1) check on changes on the blackboard goal and 2) send an async cancel to the recovery server and exit the node.
* Checking for goal updates in the RecoveryNode instead, and stop children nodes if a new goal is sent.

In my opinion, the first option is more flexible, and affects a lot less behaviours than the latter.
What do you think? Do you have some other ideass of what could be done?

Reply:
The current implementation works because when we call `halt`, it sends an async cancel goal command to all the active servers. The controller doesn't get any update from preemption, nor should it, so that the preemption doesnt act like a cancel and stop the robot in its tracks. 

The planner does not know about it, you are correct, and that's something I'm saying we should change by implementing the preemption command, propogate that to all the BTActionNode types, which includes the planner. The controller should just pass since it should keep on keeping on. 

e.g. I think all you need to do is create this on_preemption() method, which for the recoveries sends a cancel and for the planner sends the new goal. The controller is just taking the default behavior of empty method. That seems to be a general solution. Agree? If so, then the recoveries can then just derive from some BT recovery action node class which implements the on_preemption() method for cancelling and then we change that to be what all the recoveries derive from so that there's no duplication and all recoveries act the same way.

Reply:
Yes I think we are on the same page now. Any ideas about how to trigger the on_preemption() method? I realised after your comments that the blackboard flag won't work since multiple nodes might be running at the same time. Should it be done by checking for a change in the goal? Or is there any methods on the tree able to notify something like this across the whole tree? (something similar to halt that can be propagated through the tree)

Reply:
Perhaps https://github.com/ros-planning/navigation2/blob/master/nav2_behavior_tree/include/nav2_behavior_tree/behavior_tree_engine.hpp#L48

Look if its running (e.g. not idle and not success / failed) and if so, call this? @facontidavide might have an opinion on this; How to call preemption on all running nodes in the tree. But seems reasonable to use a lambda to do this as a visitor. Unless its high overhead, slow, or we can find a list of running nodes somehow without traversing.

Reply:
Okay, I'll take a look at that recursive visitor, and update the pr if I make it work. Thanks!

Reply:
I have done some research on what to do with the recursive visitor. This is what I got, tell me if you have some other ideas (my c++ skills are stretching the limit with this one haha):

* We can use the function ```BT::applyRecursiveVisitor``` to go through all nodes.
* We then pass a lambda function to the recursive visitor looking like:
```
auto notify_preemption = [](BT::TreeNode * node) {    
BtActionNode * action_node = dynamic_cast<BtActionNode>(node);
    if (action_node) {
      action_node->on_preemption();
    }
};
```
* That would be great if it wasn't because the BtActionNode is a template class, meaning that cannot be casted, which means that to go through all action nodes we need to define specific lambdas for each action -> a nonsense.

What can we do? To fix this the only thing I can come with would be creating a intermidiate class between the `BT::ActionNodeBase` and our `BtActionNode` that encapsulates the functionalities of `BtActionNode` that are not action-specific. But that starts sounding like a good refactor of the behavior_tree package already..

Any Ideas?

Reply:
https://github.com/BehaviorTree/BehaviorTree.CPP/blob/master/include/behaviortree_cpp_v3/action_node.h#L35

It looks like the ActionNodes have a type method, so you can check if this type is an action node (actually, all nodes have a type, so you can use that to find the action nodes). That should get you past the type cast. 

I think to be able to call it, we have a couple of options:
- Add a new method to the `TreeNode` class in BT.CPP for preemption-like behavior.
- Add a shim class between `BT::ActionNodeBase` and our `BtActionNode<>` that defines all our virtual interface (`on_XYZ()`, on_preemption, on_tick, on_...). Then our `BtActionNode<>` inherits from that, and we can type cast into that non-templated class
- Reimplement the `halt()` methods to do something clever. 
- Probably a fourth option I'm missing.  

In that lambda, you should also check if its running, `node->status() == BT::Running`, don't do preemption on things not actually running. 

I'm trying to keep a high level mindset and not get distracted by details: We want the ability for each BTActionNode to be able to be cancelled (halt) or preempted (overrides with new goal for planner/controller/everything else, cancelled for recoveries as a special case).

Reply:
I'll try to do this one, because I think it's the one that will look clearest without changing anything in bt.cpp:
> Add a shim class between BT::ActionNodeBase and our BtActionNode<> that defines all our virtual interface (on_XYZ(), on_preemption, on_tick, on_...). Then our BtActionNode<> inherits from that, and we can type cast into that non-templated class. If its just the pure virtual interfaces, its not really much of a refactor.

> In that lambda, you should also check if its running, node->status() == BT::Running, don't do preemption on things not actually running.

Thanks for the point, will do :)

Reply:
Got it, overall, this is a pretty big change, so also think about how you can write a couple unit tests to make sure this works. In particular: in the cases where multiple or single nodes are running how this will work. Both as a verification that this works and also so that when people mess around with the BT code in the future it doesn't cause a regression.

Reply:
Hi, any update?

Reply:
Hi, I fixed it as we said, with the shim class for the dynamic cast.
I'm working on the tests, but if you want to take a look, I can push the changes I have already.
There is still an issue that I'm not sure how to fix, and that is that if there is list of recovery actions under a control node, let's say wait -> spin, the preemption can cancel the wait, but the control node will jump into the spin. To fix this we would need to create specific control nodes, and maybe an extra status, apart from idle, running, success and failure, I'm not sure.

Reply:
Mhm. Maybe get the type and do the on_preemption for the action nodes but then for the control flow nodes, we cancel them or reset state... mhm... I'm not quite sure that generalizes in case there's a recovery and controller under the same control flow tree

Reply:
Yes, it gets more complex the more you look at the tree, because it's difficult to generalize on where you should stop stopping the parent nodes.
Maybe we can add a on_preemption to the recovery control nodes as well so they stop the second child (which is the recovery behavior) if they are running the recovery?
That way we can add a "stops child2 and runs child1 when restarted"?

Reply:
Ugh, that sounds less than ideal. But I don't know right now what to do here without having to modify all of the control flow nodes, which its sounding like we need to do.

I really don't like that since we're digging ourselves into a hole with these really niche custom control flow nodes that don't have any basis in literature. If anyone else implements a BT using this, its not going to be clear all of these weird things we had to do

Reply:
Its making me wonder if we should actually halt the BT on navigation preemption or something and just "don't tell" the controller about it so it continues to drive around until the new goal gets to it.

Reply:
Actually, that would be a good way of restarting the tree.
But I think we would need two separate functions to call on the tree no matter what. We can call halt in the whole tree and override the halt on the controller so it ignores it, but we would need to be able to cancel the controller as well.
Maybe it would make sense be to have a reset call on the node tree (all the way down in BT.CPP)?
Then there could be different behaviors for when a tree is completely stopped and when it is only restarted.

Reply:
> I really don't like that since we're digging ourselves into a hole with these really niche custom control flow nodes that don't have any basis in literature. If anyone else implements a BT using this, its not going to be clear all of these weird things we had to do

That's what scares me to do too.. I can't really see a good way of doing this without defining the specific nodes that work for navigation and documenting well how to use them, create some new ones and what are the limitations.

Reply:
Yea, we'd still need 2 different ones for a cancel vs preempt. Is the controller the only thing we need to keep running for a preemption? I think it is. 

What would a reset method all the way down the BT node level do? I don't see immediately why that's necessary if we have the on_preemption method. 

A problem that brings up is what to do with the controller if it doesnt get a new plan soon or if the controller fails because of the environment while this preemption-halting-stuff is happening

Reply:
Sorry, what I meant with the reset method is that it would probably be more helpfult to have a on_preemption or reset or sth method in the TreeNode base class so that it can be called from all the nodes.. This is basically the same you mentioned previously:
> Add a new method to the TreeNode class in BT.CPP for preemption-like behavior. Halt is like a tree-cancel, I think there should be a tree-preempt: exit(), or something.

And you are right, it would not be a lot different from what we are doing now. To make a difference, all nodes should have an implementation of it, and that doesn't make any sense now that I put it on words.

About handling errors on the controller on preemption, I think we can't really do much about it. With or without the preemption call, the planner processes the environment periodically, so adding a new goal and any changes in the environment could potentially affect it. If this is an issue we would need to address, I would definetely take it to a new ticket.

Reply:
So what do you suggest at this juncture, specifically? Should we be submitting a PR to BT.CPP to add a method we can use for preemption (some name that is reasonable for BT library)? Or what? 

I don't think we can just ignore the controller... the more I think about it, the less that that line of thinking works in my mind. If you had multiple controller servers and you preempted to another part of the tree to a different controller server than all of this falls apart and you have multiple controllers fighting to send commands.

Reply:
I really don't know what's the best option here.

I see the issue you are thinking about the controller now. That invalidates the option of restarting the whole tree on preemption. Then we need to specify behaviours for preemption.

I guess that then we could think on having a "base" node for all the different nodes (control, condition, action, etc), and recomend users to use subclasses of these nodes. We can then write navigation nodes with the different behaviors which we know are typically used, with the added functionality of the on_preemption behavior. We could extend as well the node status list to include a new status (maybe more in the future), to make a distintion between canceled and failed (doing this would allow more complex behaviours of the control nodes, like return failure if a child fails, but return success if the child returns cancelled).

What do you think? The more I think about it, the more it makes sense in my head to subclass the different nodes from BT.CPP to use specifically on navigation.
What about creating a template where we basically add the on_preemption method to all nodes? (that would probably end up pretty bad haha at least I can't wrap it around my head)

Reply:
Sounds good to me, do it. Also make sure you document some of this. We're going to need to explain this in readme files, in migration guides, and developer guides.

Reply:
Frankly, maybe you should clear your head and start at this from another angle. This is getting a little wild for what should be relatively simple.

Reply:
I also think that it should be simple, however I think that we can agree from our conversation that it's not so easy with the current implementation (none of our suggestions seem to be a good approach to it).

Yes, I'll think it over again. I am getting to know very well the behaviour tree at the end. Maybe there is some option I'm missing on BT.CPP and that would allow us to fix this in a second.

Reply:
Or perhaps if there’s some “one-shot” solution in BT.CPP we could add in a new virtual method like halt for our needs. I don’t recall from our discussions if that would actually just solve everything. I think Davide would be open to it if its named well & has some other use for other people.

Reply:
Is it correct to say that the only thing right now that’s not acting properly on preemption is the recoveries? Or also planner? Mhm... ok so all action nodes... and even the throttle decorator since it should let it replan immediately... 

This preemption thing is a pain.

Reply:
This issue kind of addresses our problem: https://github.com/BehaviorTree/BehaviorTree.CPP/issues/106

There it's mentioned that the `CoroActionNode` is useful for preemptions, however, I do not see any methods in other nodes able to act differently with this kind of node, meaning that for us it's the same.

We mainly want to be able to stop all recoveries when a new goal is sent. For the other nodes, I believe it is okay the way it is (planner and controller work as expected atm if the proper behaviour tree is built). This means that we have a tree of the shape:
<img src=https://user-images.githubusercontent.com/14293015/80463995-7891ea00-897c-11ea-99c0-b39c41fad87a.png width="300">
We have a root node that switches between two behaviours: SequenceStar and a main action.
We want to be able to cancel actions a, b and c no matter which of them is running and make the parent node return success, so the main action can start running.

To recap on what we talked so far:
* We thought of using an ActionNode-specific method that we can call on preemption of the global goal, but there is a problem with the tree: Even if we stop the running action node, it's parent node is not stopped, it just jumps to the next action.
* We thought of resetting the whole tree by somehow halting it without stopping the controller, but that implementation would fail if more than one controller is in use.
* We discussed modifying the recovery control node to halt its second child on preemption. I.e., halt the recovery behaviour branch.
* We also thought of using the blackboard to trigger the necessary actions, but that is discarded since there is no way of knowing how many nodes are running at the same time.
* We cannot force a success with a decorator to the SequenceStar output since we want to be able to percive an error on the recovery actions (this could be useful if possible, as we could simply return failure from action nodes a, b or c and that would stop the recovery).

Maybe @facontidavide knows how we can fix this on a way that makes sense.

Reply:
If we had a `halt` counterpart in the tree node, couldn't we override it for control flow nodes and mess with the internal state to return success / failure? We could set a flag and next time it's ticked, we just return some value set by the flag. Its not ideal, but it could work (?) though it doesn't set the state immediately, its only after its reticked in the new preempted goal request.

Reply:
I'm not sure I understand how that would work, but would we need to override the control nodes that we want to use then?

> Please look over my comment on that ticket and make sure it makes sense to you / is correct

Thanks for commenting, I agree with your description, specially what you mentioned about adding the communication child to parent (I believe that would be something useful for the behavior tree itself and not just us). Let's continue this discussion in that thread, I'm sure Davide will know what can be done.